ehrql/serializer.py

```diff
@@ -0,0 +1,265 @@
+import datetime
+import functools
+import json
+import pathlib
+
+from ehrql.codes import BaseCode
+from ehrql.file_formats.base import BaseDatasetReader
+from ehrql.measures.measures import Measure
+from ehrql.query_language import DummyDataConfig
+from ehrql.query_model.column_specs import ColumnSpec
+from ehrql.query_model.nodes import Node, Position, Value
+from ehrql.query_model.table_schema import BaseConstraint, Column, TableSchema
+from ehrql.utils.module_utils import get_all_subclasses
+
+
+# Mapping of names to types for all types that we support serializing/deserializing
+TYPE_REGISTRY = {
+    type_.__qualname__: type_
+    for type_ in [
+        bool,
+        int,
+        float,
+        str,
+        datetime.date,
+        tuple,
+        dict,
+        frozenset,
+        Position,
+        TableSchema,
+        Column,
+        pathlib.Path,
+        ColumnSpec,
+        DummyDataConfig,
+        Measure,
+        *get_all_subclasses(Node),
+        *get_all_subclasses(BaseCode),
+        *get_all_subclasses(BaseConstraint),
+        *get_all_subclasses(BaseDatasetReader),
+    ]
+}
+
+TYPE_REGISTRY_INVERSE = {type_: name for name, type_ in TYPE_REGISTRY.items()}
+
+# Ensure no name clashes
+assert len(TYPE_REGISTRY) == len(TYPE_REGISTRY_INVERSE)
+
+
+def type_name(obj):
+    return TYPE_REGISTRY_INVERSE[type(obj)]
+
+
+def serialize(value):
+    return json.dumps(Marshaller.to_dict(value), indent=2)
+
+
+def deserialize(data):
+    return Unmarshaller.from_dict(json.loads(data))
+
+
+class Marshaller:
```

Comment:
That's fair. I spent quite a lot of time staring at outputs while building this but none of that made it in to the test suite. I'll see if I can add something helpful.

Reply:
FWIW, I felt less of a need for this once I'd actually read the code, so there isn't a fundamental comprehensibility problem. But it would be more reader-friendly to have an overview before diving in.